packages/builders/src/components/v2/Section.ts

```diff
@@ -0,0 +1,206 @@
+import type {
+	APITextDisplayComponent,
+	APISectionComponent,
+	APIButtonComponentWithCustomId,
+	APIThumbnailComponent,
+	APIButtonComponentWithSKUId,
+	APIButtonComponentWithURL,
+} from 'discord-api-types/v10';
+import { ComponentType } from 'discord-api-types/v10';
+import { normalizeArray, type RestOrArray } from '../../util/normalizeArray.js';
+import { resolveBuilder } from '../../util/resolveBuilder.js';
+import { validate } from '../../util/validation.js';
+import { ComponentBuilder } from '../Component.js';
+import { resolveAccessoryComponent, type ButtonBuilder } from '../Components.js';
+import {
+	DangerButtonBuilder,
+	PrimaryButtonBuilder,
+	SecondaryButtonBuilder,
+	SuccessButtonBuilder,
+} from '../button/CustomIdButton.js';
+import { LinkButtonBuilder } from '../button/LinkButton.js';
+import { PremiumButtonBuilder } from '../button/PremiumButton.js';
+import { sectionPredicate } from './Assertions.js';
+import { TextDisplayBuilder } from './TextDisplay.js';
+import { ThumbnailBuilder } from './Thumbnail.js';
+
+export type SectionBuilderAccessory = ButtonBuilder | ThumbnailBuilder;
+
+export interface SectionBuilderData extends Partial<Omit<APISectionComponent, 'accessory' | 'components'>> {
+	accessory: SectionBuilderAccessory;
+	components: TextDisplayBuilder[];
+}
+
+export class SectionBuilder extends ComponentBuilder<APISectionComponent> {
+	private readonly data: SectionBuilderData;
+
+	public get components(): readonly TextDisplayBuilder[] {
+		return this.data.components;
+	}
+
+	public constructor({ components = [], accessory, ...rest }: Partial<APISectionComponent> = {}) {
+		super();
+		this.data = {
+			...structuredClone(rest),
+			// Nasty use of `!` here, but we will validate this in toJSON anyways
```

Comment:
Yeah I don't like this. The pattern I adapt generally is making the prop optional in the `XData` interface and casting in `validate`

Reply:
I mean, I'd say "we assert something that's not true" is much less true when it's just the cast after `validate`, given it'll throw if our data doesn't match the type we cast to.